ocaml/sdk-gen/go/gen_go_helper.ml

```diff
@@ -0,0 +1,167 @@
+    | DateTime ->
+        "time.Time"
+    | Enum (name, _kv) ->
+        snake_to_camel name
+    | Set ty ->
+        "[]" ^ string_of_ty ty
+    | Map (ty1, ty2) ->
+        let s1 = string_of_ty ty1 in
+        let s2 = string_of_ty ty2 in
+        "map[" ^ s1 ^ "]" ^ s2
+    | Ref r ->
+        snake_to_camel r ^ "Ref"
+    | Record r ->
+        snake_to_camel r ^ "Record"
+    | Option ty ->
+        string_of_ty ty
+
+  let fields_of_obj obj =
+    let rec flatten_contents contents =
+      List.fold_left
+        (fun l -> function
+          | Field f ->
+              f :: l
+          | Namespace (_name, contents) ->
+              flatten_contents contents @ l
+        )
+        [] contents
+    in
+    let fields = flatten_contents obj.contents in
+    let concat_and_convert field =
+      let concated =
+        String.concat "" (List.map snake_to_camel field.full_name)
+      in
+      match concated with
+      | "Uuid" | "Id" ->
+          String.uppercase_ascii concated
+      | _ ->
+          concated
+    in
+    List.map
+      (fun field ->
+        let ty = string_of_ty field.ty in
+        `O
+          [
+            ("name", `String (concat_and_convert field))
+          ; ("description", `String (String.trim field.field_description))
+          ; ("type", `String ty)
+          ]
+      )
+      fields
+
+  let xenapi objs =
+    List.map
+      (fun obj ->
+        let fields = fields_of_obj obj in
+        let event_snapshot =
+          if String.lowercase_ascii obj.name = "event" then
+            [
+              `O
+                [
+                  ("name", `String "Snapshot")
+                ; ( "description"
+                  , `String
+                      "The record of the database object that was added, \
+                       changed or deleted"
+                  )
+                ; ("type", `String "RecordInterface")
+                ]
+            ]
+          else
+            []
+        in
+        let obj_name = snake_to_camel obj.name in
+        let event_session_value = function
+          | "event" ->
+              [("event", `Bool true); ("session", `Null)]
+          | "session" ->
+              [("event", `Null); ("session", `Bool true)]
+          | _ ->
+              [("event", `Null); ("session", `Null)]
+        in
+        let base_assoc_list =
+          [
+            ("name", `String obj_name)
+          ; ("description", `String (String.trim obj.description))
+          ; ("fields", `A (event_snapshot @ fields))
+          ]
+        in
+        let assoc_list = event_session_value obj.name @ base_assoc_list in
+        (String.lowercase_ascii obj.name, `O assoc_list)
+      )
+      objs
+end
+
+let objects =
+  let api = Datamodel.all_api in
+  (* Add all implicit messages *)
+  let api = add_implicit_messages api in
+  (* Only include messages that are visible to a XenAPI client *)
+  let api = filter (fun _ -> true) (fun _ -> true) on_client_side api in
```

Comment:
There are only two diffs we don't do filters below:
`List.mem "closed" field.release.internal`  and `List.mem "closed" msg.msg_release.internal`.

Firstly, we could check the code how build the `message` in `datamodel_common.ml`. In the `call` function:
```ocaml
let call ...... =
 .....
  let call_release =
    {
      internal=
        ( match get_published lifecycle with
        | Some published ->
            get_product_releases published
        | None ->
            ["closed"]
        )
    ; opensource= get_oss_releases in_oss_since
    ; internal_deprecated_since= get_deprecated lifecycle
    }
  in
  {
 ......
  ; msg_release= call_release
 ......
  }
``` 
and the definition of `get_product_releases` is:
```ocaml
let get_product_releases in_product_since =
  let rec go_through_release_order rs =
    match rs with
    | [] ->
        (* numbered release *)
        ["closed"; in_product_since]
    | x :: xs when code_name_of_release x = in_product_since ->
        "closed" :: in_product_since :: List.map code_name_of_release xs
    | _ :: xs ->
        go_through_release_order xs
  in
  go_through_release_order release_order
``` 
We could see from the code every `msg_release` has "closed".

Then let's check how we build field in function `field`:
```ocaml
(** Make an object field record *)
let field ...... =
  ......
  let release =
    {
      internal=
        ( match get_published lifecycle with
        | Some published ->
            get_product_releases published
        | None ->
            ["closed"]
        )
    ; opensource= get_oss_releases in_oss_since
    ; internal_deprecated_since= get_deprecated lifecycle
    }
  in
  Field
    {
      release
    ......
    }
``` 
So every `field.release` has "closed" too.

Since all of them has "closed", we don't need filter `List.mem "closed" field.release.internal`  and `List.mem "closed" msg.msg_release.internal` here.
And it doesn't include more API than other languages.

Reply:
Can this value be moved to the common library and be shared among all the languages? I'd like to see a consistent way to generate the languages' view. If that isn't possible, at least this will highlight issues in certain languages, or problems with the datamodel, like you've seen.

The problem about `closed` means internal is currently useless, I'm not sure if it should be dropped completely, or it can serve a function, @robhoes might know more, but he won't be able to respond for a few days. I don't we should solve the prolem here

Reply:
This part seems the same as the one in `ocaml/idl/json_backend/gen_json.ml`.
Maybe they could be consolidated as a common part?

Reply:
Yeah, I copied it from `gen_json.ml`.
I'd like to put it in `ocaml/sdk-gen/common/CommonFunctions.ml`.
Some languages could reuse the function, java and C#.